compiler/rustc_borrowck/src/region_infer/mod.rs

```diff
@@ -2168,17 +2175,38 @@ impl<'tcx> RegionInferenceContext<'tcx> {
                 }
             }
 
+            // If an "outlives 'static" constraint was from use as a const or static, blame that.
+            if target_is_static
+                && blame_source
+                && let Some(old_best) = categorized_path.iter().min_by_key(|p| p.category)
+                && matches!(
+                    old_best.category,
+                    ConstraintCategory::UseAsConst
+                        | ConstraintCategory::UseAsStatic
+                        | ConstraintCategory::Cast {
+                            is_implicit_coercion: true,
+                            unsize_to: Some(_)
+                        }
+                )
+            {
+                // FIXME(dianne): `BorrowExplanation::add_object_lifetime_default_note` depends on a
+                // coercion being blamed, so revert to the old blaming logic to prioritize that.
+                // The note's logic should be reworked, though; it's flaky (#131008 doesn't have a
+                // coercion, and even with this hack, one isn't always blamed when present).
+                // Only checking for a coercion also makes the note appear where it shouldn't
+                // shouldn't (e.g. `tests/ui/borrowck/two-phase-surprise-no-conflict.stderr`).
```

Comment:
from what I can tell, this FIXME is stating that this branch is causing us to not emit a `add_object_lifetime_default_note` in cases where we should. Please link to an affected test

Reply:
ah, the wording of the FIXME is unclear; I'll see if I can make it better. I've made sure to still emit that note where it was previously emitted. that branch is a hack specifically to keep the note present in [tests/ui/traits/trait-object-lifetime-default-note.stderr](https://github.com/dianne/rust/blob/better-blame-constraints-for-static/tests/ui/traits/trait-object-lifetime-default-note.stderr), since otherwise the `CallArgument` constraint would be blamed and the note would disappear; it'll be more apparent once I get rid of the `UseAsConst`/`UseAsStatic` special-casing and split the `Cast` case into a separate commit. I'll refer to that test in the FIXME too.

the missing notes the FIXME is referring to are cases where the note would make sense (and imo should be emitted)[^1], but it wasn't emitted even before this change. the note depends on a coercion being blamed, but it won't always be blamed for region errors from object lifetime defaults (nor will a coercion necessarily be present). handling that properly is (afaict) nontrivial; I'd be happy to do what I can there in a separate PR, but I might either need some help with wording or some direction on what heuristics would make sense, since doing it as perfectly as possible seems quite involved.

[^1]: see #131008, which would probably need a fairly general solution. I remember finding in-tree tests that are missing the note too though (which would require a less-general but still nontrivial fix), so I can dig those up if it'd help